R/utils.R

```diff
@@ -169,77 +169,94 @@ parse_labels <- function(string, return_vector = FALSE) {
-#'
-#' @importFrom rlang .data
-#' @keywords internal
-
-checkbox_appender <- function(field_name, string) {
-  prefix <- paste0(field_name, "___")
-
-  out <- parse_labels(string)
-  out$raw <- tolower(out$raw)
-  # append each element of the split vector with the field_name prefix and then
-  # recombine
-  out <- paste0(prefix, out[[1]])
-
-  out
-}
-
 #' @title
 #' Update metadata field names for checkbox handling
 #'
 #' @description
-#' Takes a \code{db_metadata} object and appends a \code{field_name_updated}
-#' field to the end for checkbox variable handling.
-#'
-#' @returns Column \code{field_name_updated} appended to a given REDCap
-#' \code{db_metadata} object.
+#' Takes a \code{db_metadata} object and:
+#' \itemize{
+#'   \item replaces checkbox field rows with a set of rows, one for each
+#'   checkbox option
+#'   \item appends a \code{field_name_updated} field to the end for checkbox
+#'   variable handling
+#'   \item updates \code{field_label} for any new checkbox rows to include the
+#'   specific option in parenthese
+#' }
+#'
+#' @returns Column \code{db_metadata} with \code{field_name_updated} appended
+#' and \code{field_label} updated for new rows corresponding to checkbox options
 #'
 #' @param db_metadata The REDCap metadata output defined by
 #' \code{REDCapR::redcap_metadata_read()$data}
 #'
-#' @return Returns an updated REDCap metadata object with updated field names
+#' @details
+#' Assumes \code{db_metadata}:
+#' \itemize{
+#'   \item has non-zero number of rows
+#'   \item contains \code{field_name} and \code{field_label} columns
+#' }
 #'
 #' @importFrom tidyr unnest
+#' @importFrom dplyr %>% select mutate
+#' @importFrom tibble tibble
+#' @importFrom rlang .data
+#' @importFrom stringr str_replace_all str_trim str_squish
 #'
 #' @keywords internal
 
 update_field_names <- function(db_metadata) {
   out <- db_metadata %>%
     mutate(
-      field_name_updated = list(c())
+      updated_metadata = list(c())
     )
 
-  # Apply checkbox appender function to rows of field_type == "checkbox"
-  # Assign all field names, including expanded checkbox variables, to a list col
   for (i in seq_len(nrow(out))) {
     if (out$field_type[i] == "checkbox") {
-      out$field_name_updated[i] <- list(
-        checkbox_appender(field_name = out$field_name[i],
-                          string = out$select_choices_or_calculations[i])
+      # If checkbox field, parse labels fill updated_metadata with a tibble
+      # containing updated field names and labels
+      parsed_labs <- parse_labels(out$select_choices_or_calculations[i])
+
+      # Build updated field names and labs
+      clean_names <- paste(
+        out$field_name[[i]],
+        tolower(parsed_labs$raw),
+        sep = "___"
+      )
+
+      # Strip embedding logic and white space before appending to field labels
+      clean_labs <- parsed_labs$label %>%
+        # TODO: replace with formatting functions?
+        str_replace_all("\\{.*?\\}", "") %>%
```

Comment:
Did we confirm that this should be a default enforced option? Is there any reason we may want to have embedding/{} available, or added to the default `formatter` options? In my mind if we're not defaulting stripping HTML tags then we shouldn't default stripping embedded fields either.

Reply:
I think @skadauke and I hastily made that decision while meeting yesterday but I don't have strong opinions. If we chose not to edit the checkbox option labels on the way into the metadata the user would still be able to do things like strip html/embedding logic on the way into `labelled` labels when they call `make_labelled()`.

Reply:
I'm could see going either way since they serve different purposes, but I tend to lean toward showing the user everything and then letting them decide what they want to take away (or giving them the option to do so).

Reply:
To clarify - we're not talking about formatting labels for `labelled` functionality here but about changing the label of an option of a radio/checkbox field (although this also ultimately affects the tibble's variable labels). I believe that you _always_ want to remove field embedding code here.

Reply:
Correct. The question is what processing we should do to create the contents of `supertbl$redcap_metadata` before `labelled` gets involved

Reply:
It's a little bit confusing, but now that I'm thinking about it more there I believe that there are two different scenarios where field embedding code might pop up and they may need to be handled differently.

The first is a multi-checkbox such as in the example above. Here we'd want `Other {crs_anti_tcell_therapy_types_specify}` to be `Other`, and if this was the only scenario this could reasonably be handled with a formatter function.

The second scenario in which field embeddings can come up is in the context of a radio field (and possibly also dropdown but I'm not sure). In this case, `Other {something_specify}` doesn't affect the `labelled` variable field. But it _does_ affect the factor label of the variable which will be `Other {something_specify}` instead of `Other`.

Reply:
Should we just give the user the option of applying a formatting function to all checkbox/radio/dropdown labels as an argument in `read_redcap_tidy()` or is that overly complex?

If yes we could centralize the logic in `parse_labels()` so that it would flow into both the metadata field labels (for checkboxes) and the factor levels (for everything else).

Reply:
I'm okay with this. I kind of agree with @rsh52's suggestion above https://github.com/CHOP-CGTDataOps/REDCapTidieR/pull/73#discussion_r1020283174 (which he may or may not still stand by) that if we're stripping embedding logic we also strip html.

I can't think of a reason a user would want raw html in their factor levels

Reply:
Yea, it's still my opinion that applying these changes should be done synonymously even if they serve different purposes. I could see the same person that says "I want to see the HTML" also saying "I want to see the field embedding". It may not be the most useful thing in 99% of cases but it's still a true representation of the output and may have its place somewhere. Some REDCap aficionado may be confused when things they expect to be there no longer are.